model/model.py

```diff
@@ -0,0 +1,170 @@
+from typing import List, Optional, Self
+
+from torch import nn
+from torch.utils.data import DataLoader
+import torch
+import snntorch as snn
+import pandas as pd
+
+from datasets.src.zenke_2a.constants import TRAIN_DATA_PATH
+from datasets.src.zenke_2a.constants import TEST_DATA_PATH
+from datasets.src.zenke_2a.dataset import SequentialDataset
+
+
+class Settings:
+    def __init__(self,
+                 layer_sizes: list[int],
+                 beta: float,
+                 learn_beta: bool,
+                 num_steps: int,
+                 data_size: int,
+                 batch_size: int,
+                 learning_rate: float,
+                 epochs: int) -> None:
+        self.layer_sizes = layer_sizes
+        self.beta = beta
+        self.learn_beta = learn_beta
+        self.num_steps = num_steps
+        self.data_size = data_size
+        self.batch_size = batch_size
+        self.learning_rate = learning_rate
+        self.epochs = epochs
+
+
+class LayerSettings:
+    def __init__(self, prev_size: int, size: int, next_size: int, beta: float, learn_beta: bool,
+                 batch_size: int, learning_rate: float) -> None:
+        self.prev_size = prev_size
+        self.size = size
+        self.next_size = next_size
+        self.beta = beta
+        self.learn_beta = learn_beta
+        self.batch_size = batch_size
+        self.learning_rate = learning_rate
+
+
+class Layer(nn.Module):
+    def __init__(self, layer_settings: LayerSettings) -> None:
+        super().__init__()
+
+        # weights from prev layer to this layer
+        self.forward_weights = nn.Linear(
+            layer_settings.prev_size, layer_settings.size)
+        # TODO: how does it know the size?
+        self.forward_lif = snn.Leaky(
```

Comment:
What is the usecase for using snn.LeakyParallel?

Reply:
Pros: linear weights included
Cons: all timesteps passed at once

IIUC we need each timestep to be trained in online fashion. So that means we would need to keep it like this. This approach seems more flexible also as it exposes membrane potential if we ever need that.

I forgot about this TODO though I need to understand this part better. Not sure what is happening right now. I need to confirm LIF is inferring the size correctly from the `current` in the forward pass.